pkg/ipmi/ipmi.go

```diff
@@ -0,0 +1,191 @@
+// Copyright 2019 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package ipmi
+
+import (
+	"fmt"
+	"os"
+	"syscall"
+	"unsafe"
+)
+
+const (
+	_IOC_WRITE = 0x1
+	_IOC_READ  = 0x2
+
+	_IOC_NRBITS   = 8
+	_IOC_TYPEBITS = 8
+	_IOC_SIZEBITS = 14
+	_IOC_NRSHIFT  = 0
+
+	_IOC_TYPESHIFT = _IOC_NRSHIFT + _IOC_NRBITS
+	_IOC_SIZESHIFT = _IOC_TYPESHIFT + _IOC_TYPEBITS
+	_IOC_DIRSHIFT  = _IOC_SIZESHIFT + _IOC_SIZEBITS
+
+	_IPMI_BMC_CHANNEL                = 0xf
+	_IPMI_BUF_SIZE                   = 1024
+	_IPMI_IOC_MAGIC                  = 'i'
+	_IPMI_NETFN_APP                  = 0x6
+	_IPMI_OPENIPMI_READ_TIMEOUT      = 15
+	_IPMI_SYSTEM_INTERFACE_ADDR_TYPE = 0x0c
+
+	_BMC_SET_WATCHDOG_TIMER = 0x24
+	_BMC_GET_WATCHDOG_TIMER = 0x25
+
+	_IPM_WATCHDOG_NO_ACTION    = 0x00
+	_IPM_WATCHDOG_SMS_OS       = 0x04
+	_IPM_WATCHDOG_CLEAR_SMS_OS = 0x10
+)
+
+var (
+	_IPMICTL_RECEIVE_MSG  = iowr(_IPMI_IOC_MAGIC, 12, int(unsafe.Sizeof(recv{})))
+	_IPMICTL_SEND_COMMAND = ior(_IPMI_IOC_MAGIC, 13, int(unsafe.Sizeof(req{})))
+)
+
+type Ipmi struct {
+	fd int
+}
+
+type msg struct {
+	netfn   byte
+	cmd     byte
+	dataLen uint16
+	data    unsafe.Pointer
+}
+
+type req struct {
+	addr    *systemInterfaceAddr
+	addrLen uint32
+	msgid   int64
+	msg     msg
+}
+
+type recv struct {
+	recvType int32
+	addr     *systemInterfaceAddr
+	addrLen  uint32
+	msgid    int64
+	msg      msg
+}
+
+type systemInterfaceAddr struct {
+	addrType int32
+	channel  int16
+	lun      byte
+}
+
+func ioc(dir int, t int, nr int, size int) int {
+	return (dir << _IOC_DIRSHIFT) | (t << _IOC_TYPESHIFT) |
+		(nr << _IOC_NRSHIFT) | (size << _IOC_SIZESHIFT)
+}
+
+func ior(t int, nr int, size int) int {
+	return ioc(_IOC_READ, t, nr, size)
+}
+
+func iowr(t int, nr int, size int) int {
+	return ioc(_IOC_READ|_IOC_WRITE, t, nr, size)
+}
+
+func ioctl(fd int, name int, data unsafe.Pointer) syscall.Errno {
+	_, _, err := syscall.RawSyscall(syscall.SYS_IOCTL, uintptr(fd), uintptr(name), uintptr(data))
+	return err
+}
+
+func fd_set(fd int, p *syscall.FdSet) {
+	p.Bits[fd/64] |= 1 << (uint(fd) % 64)
+}
+
+func Open(devnum int) (*Ipmi, error) {
+	d := fmt.Sprintf("/dev/ipmi%d", devnum)
+
+	f, err := os.OpenFile(d, os.O_RDWR, 0)
+	if err != nil {
+		return nil, err
+	}
+
+	fd := int(f.Fd())
```

Comment:
Why not keep the os.File in the struct? Then you would not need to write a Close.

Reply:
> Why not keep the os.File in the struct?

Because for the ioctl calls I need the file descriptor.

> Then you would not need to write a Close.

I didn't follow this one. Do you mean avoid writing a close method by doing something like `ipmi.File.Close()`?

Reply:
no, you can do two things:
in your struct, just embed
os.File
and then when you need to close, you can
ipmi.Close()
and when you need an fd,
ipmi.Fd()

then you don't even need to use the fd in the struct.

Reply:
Got it. I added it as an embedded type.